src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -2875,184 +3102,1418 @@ private void ensurePublishCapability(
                 }
             }
         }
-    }
-
-    private static boolean invalidWillQos(
-        int flags)
-    {
-        return (flags & WILL_QOS_MASK) == WILL_QOS_MASK;
-    }
-
-    private static boolean isSetWillQos(
-        int flags)
-    {
-        return (flags & WILL_QOS_MASK) != 0b0000_0000;
-    }
-
-    private static boolean isSetWillRetain(
-        int flags)
-    {
-        return (flags & WILL_RETAIN_MASK) != 0;
-    }
 
-    private static boolean isSetWillFlag(
-        int flags)
-    {
-        return (flags & WILL_FLAG_MASK) != 0;
-    }
+        private void doEncodeWillMessageIfNecessary(
+            MqttSessionStream sessionStream,
+            long traceId,
+            long authorization)
+        {
+            if (sessionStream != null && sessionStream.willFlagSet && sessionExpiryInterval == 0)
+            {
+                final MqttDataExFW dataEx = sessionStream.willMessage;
+                final OctetsFW payload = sessionStream.willPayload;
+                final String topic = dataEx.topic().asString();
 
-    private static boolean isSetUsername(
-        int flags)
-    {
-        return (flags & USERNAME_MASK) != 0;
-    }
+                if (willStream == null)
+                {
+                    this.willStream = new MqttWillStream(sessionStream.routeId, authorization, 0, topic);
+                    willStream.doApplicationBeginOrFlush(traceId, authorization, sessionStream.clientIdentifier, affinity, topic,
+                        NO_FLAGS, 0, PUBLISH_ONLY);
+                }
```

Comment:
Can the rest of this method be replaced by a call to `willStream.publishWillMessage(traceId)` instead?